JavaScript code generation for traits

chengluyu opened the pull request:
This PR is WIP. I'll write a summary when it's ready.

chengluyu committed:
Classes now inherit fields from direct base traits

But not indirect base traits.

chengluyu committed:
Classes now inherit fields from traits recursively

chengluyu committed:
Implement basic but incomplete trait code generation

Here is a sum up.

- Cannot `defineProperty` on non-objects (in Yicong.mls).
- Cannot redefine property (in Traits.mls).
- Should call `implement`s in class constructors.

chengluyu commented on shared/src/main/scala/mlscript/JSBackend.scala:
```diff
@@ -223,6 +226,82 @@ class JSBackend {
+        N,
+        Nil,
+        Ls(
+          const("tag", id("Object")()),
```
```suggestion
          const("tag", id("Symbol")()),
```

LPTK replied:
Didn't you want to use the `Symbol` as a key?

chengluyu replied:
Yes! I fixed it!

chengluyu committed:
Fix issues mentioned in the last meetings

- Add null test to `Trait.is`.
- Check if instance is object in `Trait.implement`.
- Class constructors implement traits now.
- Fix typo `Object()` to `Symbol()` in `Trait.implement`.

chengluyu commented:
We probably need to clean up `JSBackend.translateTraitDeclaration`.

chengluyu committed:
Merge branch 'mlscript' into codegen-traits

LPTK committed:
Add problematic test cases

chengluyu committed:
Fix trait applications on primitives

chengluyu committed:
Fix empty class body when inheriting from traits

chengluyu committed:
Inject trait members only if the field is available

chengluyu committed:
Fix traits not inheriting from their bases

chengluyu committed:
Add an additional test about trait inheritance

LPTK committed:
Forbid instantiation of traits that contain methods

LPTK reviewed (approved):
LGTM!

Do you have a test case that shows matching on a trait constructed on a primitive value?

As in:

```
t1 = T1 1
//│ t1: 1 & t1
//│   = [Number: 1]

case t1 of { T1 -> () }
//│ res: ()
//│    = []
```

You can go ahead and squash-merge.

chengluyu committed:
Add a case showing pattern matching on apply traits to primitives

chengluyu committed:
Hide trait methods (same as class methods)

chengluyu committed:
Merge branch 'mlscript' into codegen-traits
